### PR TITLE
feat(tokenizer): add fastokens backend

### PR DESCRIPTION
**Pull request by hashkanna**
## Motivation

Port the merged upstream SGLang fastokens tokenizer backend from https://github.com/sgl-project/sglang/pull/23753 into SGL-JAX.

This keeps Hugging Face tokenizers as the default while adding an opt-in `--tokenizer-backend fastokens` path for faster CPU-side tokenization on supported tokenizer families. The change targets tokenizer-manager TTFT overhead for long prompts and is intentionally narrow: no model forward path, kernel, scheduler policy, or default serving behavior changes.

No separate RFC/design doc is included because this is a small, opt-in port of an already-reviewed upstream SGLang feature rather than a new SGL-JAX architecture feature.

## Modifications

- Add optional `fastokens` extra: `fastokens>=0.1.1,<0.2.0`.
- Add `--tokenizer-backend {huggingface,fastokens}` to `ServerArgs`; default remains `huggingface`.
- Patch Transformers once, process-wide, when `fastokens` is selected.
- Return clear errors when `fastokens` is not installed or does not support the requested tokenizer.
- Thread `server_args.tokenizer_backend` through tokenizer initialization in tokenizer, detokenizer, and scheduler managers.
- Add tests for Hugging Face/fastokens token-id parity, fastokens shim usage, actionable fastokens load errors, and invalid backend validation. Fastokens-specific tests skip unless `fastokens` is installed.

## Accuracy Tests

This does not change model weights, model forward code, sampling, logits, or generated-token selection. The validation here is tokenizer token-id equivalence for a supported tokenizer.

- `PYTHONPATH=python python/.venv/bin/python -m unittest sgl_jax.test.test_hf_transformers_fastokens -v`
  - checks `Qwen/Qwen3-0.6B` Hugging Face vs fastokens token IDs on short and long prompts
  - checks `tokenizer._tokenizer` is fastokens `_TokenizerShim`
  - checks fastokens load failures give an actionable fallback message
  - checks invalid `tokenizer_backend` values are rejected
- `PYTHONPATH=python python -m unittest python.sgl_jax.test.test_hf_transformers_fastokens -v`
  - no-fastokens environment: fastokens-specific tests skip, invalid-backend validation still runs
- Smoke checked CLI parsing stores `tokenizer_backend="fastokens"`.

- TPU VM dummy-weight server startup smoke on TPU VM (`v6e-1`, `europe-west4-a`): launched `Qwen/Qwen3-0.6B` with `--load-format=dummy --device=tpu --tp-size=1 --tokenizer-backend fastokens`; `/health` returned 200 and logs showed `tokenizer_backend='fastokens'` plus `fastokens.patch_transformers` in tokenizer, detokenizer, and scheduler processes.

## Benchmarking and Profiling

Tokenizer-only microbenchmarks with `Qwen/Qwen3-0.6B`, 36,864-token prompt, 20 encode iterations after warmup:

Local development machine:

| Backend | Median encode time | Mean encode time |
| --- | ---: | ---: |
| Hugging Face | 34.889 ms | 34.925 ms |
| fastokens | 0.938 ms | 0.935 ms |

TPU VM host CPU (`v6e-1`, `europe-west4-a`):

| Backend | Median encode time | Mean encode time |
| --- | ---: | ---: |
| Hugging Face | 34.506 ms | 34.623 ms |
| fastokens | 1.301 ms | 1.311 ms |

No end-to-end TPU performance benchmark was run because this PR only adds an opt-in tokenizer backend and does not change the default path.

## Checklist

- [x] Please use English, otherwise it will be closed.
- [x] The purpose of the PR, or link existing issues this PR will resolve.
- [x] The test plan, such as providing test command.
- [x] (Optional) The necessary documentation update.

Additional checks run:

- `pre-commit run --files python/pyproject.toml python/sgl_jax/srt/hf_transformers_utils.py python/sgl_jax/srt/managers/detokenizer_manager.py python/sgl_jax/srt/managers/scheduler.py python/sgl_jax/srt/managers/tokenizer_manager.py python/sgl_jax/srt/server_args.py python/sgl_jax/test/test_hf_transformers_fastokens.py`
- `ruff check python/sgl_jax/srt/hf_transformers_utils.py python/sgl_jax/srt/server_args.py python/sgl_jax/srt/managers/tokenizer_manager.py python/sgl_jax/srt/managers/detokenizer_manager.py python/sgl_jax/srt/managers/scheduler.py python/sgl_jax/test/test_hf_transformers_fastokens.py`
- `git diff --check`
- `python/.venv/bin/python -m py_compile python/sgl_jax/srt/hf_transformers_utils.py python/sgl_jax/srt/server_args.py python/sgl_jax/srt/managers/tokenizer_manager.py python/sgl_jax/srt/managers/detokenizer_manager.py python/sgl_jax/srt/managers/scheduler.py python/sgl_jax/test/test_hf_transformers_fastokens.py`

### CODE REVIEW

**Review by alexnails — DISMISSED** [review]

**Comment by hashkanna** [comment]
@JamesBrianD Is there anything else I need to do here to get it merged? thanks!

File: python/sgl_jax/srt/hf_transformers_utils.py [hunk 1]
```diff
@@ -236,8 +287,15 @@ def get_tokenizer(
                 "or using the `--trust-remote-code` flag in the CLI."
             )
             raise RuntimeError(err_msg) from e
-        else:
-            raise e
+
+        if tokenizer_backend == "fastokens":
+            _raise_fastokens_load_error(tokenizer_name, e)
```
Comment (pengchengneo) [inline python/sgl_jax/srt/hf_transformers_utils.py]: The fastokens error check is duplicated across three except branches (TypeError, ValueError, OSError | RuntimeError). 

Upstream consolidates this into a single except Exception block. 

That's simpler to maintain and won't silently miss a new exception type added in the future.

File: python/sgl_jax/srt/hf_transformers_utils.py [hunk 2]
```diff
@@ -165,6 +166,46 @@ def get_context_length(config):
...
+        try:
+            import fastokens
+        except ImportError:
+            raise ImportError(
+                "The fastokens package is required when tokenizer_backend='fastokens'. "
+                "Install it with: pip install 'sglang-jax[fastokens]'"
+            ) from None
+
+        fastokens.patch_transformers()
+        _FASTOKENS_PATCHED = True
```
Comment (pengchengneo) [inline python/sgl_jax/srt/hf_transformers_utils.py:217]: logger.info("fastokens backend enabled - transformers patched successfully")

add a successful log here? it is just nit

File: python/pyproject.toml [hunk 3]
```diff
@@ -61,6 +64,7 @@ cpu = [
 ]
 all = [
```
Comment (pengchengneo) [inline python/pyproject.toml:65]: have you tried using 'uv sync --extra all', ensure there are no dependency conflicts ?

File: python/sgl_jax/srt/hf_transformers_utils.py (hunk 1 above)
Comment (hashkanna) [reply python/sgl_jax/srt/hf_transformers_utils.py]: Done. I have consolidated it into a single except Exception.

File: python/sgl_jax/srt/hf_transformers_utils.py (hunk 2 above)
Comment (hashkanna) [reply python/sgl_jax/srt/hf_transformers_utils.py:217]: Added the successful log. Thanks.

File: python/pyproject.toml (hunk 3 above)
Comment (hashkanna) [reply python/pyproject.toml:65]: Verified `uv sync --extra all` on the TPU VM; it resolved and installed successfully.

File: python/sgl_jax/srt/hf_transformers_utils.py [hunk 4]
```diff
@@ -250,8 +311,8 @@ def get_tokenizer(
                 "or using the `--trust-remote-code` flag in the CLI."
             )
             raise RuntimeError(err_msg) from e
-        else:
-            raise e
+
+        raise
```
Comment (pengchengneo) [inline python/sgl_jax/srt/hf_transformers_utils.py:315]: bug here?

Comment (hashkanna) [reply python/sgl_jax/srt/hf_transformers_utils.py:315]: Hi, this is intended and not a bug. This matches upstream SGLang: the public get_tokenizer wrapper catches Exception, wraps fastokens failures, and then uses bare raise for the normal backend to preserve the original exception/traceback. The bare raise only runs inside the active except block, so there's always an exception to re-raise.

Reference:
https://github.com/sgl-project/sglang/blob/ec075d8bc5ff847f43637e9f6dd0cd03f962b024/python/sglang/srt/utils/hf_transformers/tokenizer.py#L515-L523

**Review by pengchengneo — APPROVED** [review]

**Comment by pengchengneo** [comment]
LGTM
